public/pages/Main/Main.tsx

```diff
@@ -565,6 +638,8 @@ export default class Main extends Component<MainProps, MainState> {
                                       {...props}
                                       rollupService={services.rollupService}
                                       indexService={services.indexService}
+                                      dataSourceReadOnly={this.state.dataSourceReadOnly}
+                                      setDataSourceReadOnly={this.setDataSourceReadOnly}
```

Comment:
Why don't we make it as a context service?

Reply:
Earlier I did! But after a suggestion from Rama on this PR I changed it to props only. Passing through seems props better because only 2 components need this as of now whereas as a context service it would've been accessible to all the components and may lead to accidental usage in other components in future. Also, the usage of the state is just one level below the main, so we didn't have to do props drilling which we could've prevented by making it as a context service

Reply:
So it is used only for multi-steps pages which may switch the data source picker to readonly or changeable while enter next step or go back to the previous step?

Reply:
Yes Exactly Zhou :)